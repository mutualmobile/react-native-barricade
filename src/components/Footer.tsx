import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale } from '../utils';

export type FooterProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
  titleStyle: TextStyle;
};

const Footer = (props: FooterProps): JSX.Element => {
  const { onPress, title, titleStyle } = props;
  const { themeColorStyle } = useThemedColor();

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      <TouchableOpacity onPress={onPress} style={styles.footer}>
        <Text numberOfLines={1} style={[styles.title, titleStyle]}>
          {title}
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 50,
    borderTopWidth: 1,
  },
  footer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: hScale(12),
  },
  title: {
    textAlign: 'center',
    fontSize: hScale(18),
  },
});

export { Footer };
