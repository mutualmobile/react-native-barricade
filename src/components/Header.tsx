import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';

export type ButtonProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
};

export type HeaderProps = {
  headerLeft?: ButtonProps;
  headerRight?: ButtonProps;
  title: string;
};

const Header = (props: HeaderProps): JSX.Element => {
  const { headerLeft, headerRight, title } = props;
  const { themeColorStyle } = useThemedColor();

  const renderButton = (buttonProps: ButtonProps, style: ViewStyle) => {
    return (
      <TouchableOpacity onPress={buttonProps.onPress} style={style}>
        <Text style={[styles.buttonText, themeColorStyle.primary]}>
          {buttonProps.title}
        </Text>
      </TouchableOpacity>
    );
  };

  const renderHeader = () => {
    return (
      <View style={styles.header}>
        <Text
          numberOfLines={1}
          style={[styles.title, themeColorStyle.textDark]}>
          {title}
        </Text>
        {headerLeft && renderButton(headerLeft, styles.headerLeft)}
        <View style={styles.flex} />
        {headerRight && renderButton(headerRight, styles.headerRight)}
      </View>
    );
  };

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      {renderHeader()}
    </View>
  );
};

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    borderBottomWidth: 1,
  },
  header: {
    marginHorizontal: hScale(12),
    height: vScale(50),
    alignItems: 'center',
    flexDirection: 'row',
  },
  title: {
    position: 'absolute',
    textAlign: 'center',
    fontSize: hScale(20),
    fontWeight: 'bold',
    marginHorizontal: hScale(60),
    left: 0,
    right: 0,
  },
  buttonText: {
    fontSize: hScale(18),
    margin: hScale(8),
    textAlign: 'center',
  },
  headerLeft: {
    marginRight: hScale(10),
  },
  headerRight: {
    marginLeft: hScale(10),
  },
});

export { Header };
