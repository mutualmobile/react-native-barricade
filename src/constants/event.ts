export const EventType = {
  ShowBarricadeView: 'ShowBarricadeView',
};
