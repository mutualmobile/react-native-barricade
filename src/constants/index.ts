export * from './event';
export * from './strings';
export * from './unicode';
