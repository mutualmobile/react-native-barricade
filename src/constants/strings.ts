export const Strings = {
  Back: 'Back',
  Barricade: 'Barricade',
  DisableBarricade: 'Disable Barricade',
  Done: 'Done',
  EnableBarricade: 'Enable Barricade',
  Reset: 'Reset',
  Space: ' ',
};
