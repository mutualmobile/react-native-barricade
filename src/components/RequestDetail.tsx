import React from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import {
  Barricade,
  RequestConfig,
  RequestConfigForLib,
  ResponseHandlerForLib,
} from '../network';
import { useThemedColor } from '../theme';
import { Header } from './Header';

const CheckMark = '\u2713';
const ChevronLeft = '\u2039';

const RequestDetail = ({
  barricade,
  detailData,
  onBackPressed,
}: {
  barricade: Barricade | undefined;
  detailData?: RequestConfigForLib;
  onBackPressed: () => void;
}): JSX.Element => {
  const { themeColorStyle } = useThemedColor();

  const onDeailItemPressed = (item: ResponseHandlerForLib) => {
    if (!item.isSelected) {
      detailData?.responseHandler.map(responseHandler => {
        responseHandler.isSelected = responseHandler.label === item.label;
        return responseHandler;
      });
      const requestConfig = barricade?.requestConfig.map<RequestConfig>(
        config => {
          if (config.label === detailData?.label) {
            config = detailData;
          }
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { selectedResponseLabel, ...result } = config;
          return result;
        },
      );
      if (requestConfig) {
        barricade?.updateRequestConfig(requestConfig);
      }
      onBackPressed();
    }
  };

  const renderDetailItem = ({
    item,
  }: ListRenderItemInfo<ResponseHandlerForLib>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onDeailItemPressed(item)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        {item.isSelected && (
          <Text style={[styles.icon, themeColorStyle.primary]}>
            {CheckMark}
          </Text>
        )}
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{ title: ChevronLeft + ' Back', onPress: onBackPressed }}
        title={detailData?.label ?? ''}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={detailData?.responseHandler}
        renderItem={renderDetailItem}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 20,
    paddingVertical: 12,
    borderBottomWidth: 1,
  },
  label: {
    fontSize: 18,
    fontWeight: '400',
  },
  icon: {
    fontSize: 18,
  },
});

export { RequestDetail };
