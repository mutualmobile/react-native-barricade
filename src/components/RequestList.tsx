import React, { useState } from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { Strings, Unicode } from '../constants';
import { Barricade, RequestConfigForLib } from '../network';
import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';
import { Footer } from './Footer';
import { Header } from './Header';

type RequestListProps = {
  barricade: Barricade;
  onDonePressed?: (event: NativeSyntheticEvent<any>) => void;
  onListItemPressed: (index: number) => void;
};

const RequestList = (props: RequestListProps): JSX.Element => {
  const { barricade, onDonePressed, onListItemPressed } = props;
  const { themeColorStyle } = useThemedColor();
  const [refreshList, setRefreshList] = useState(0);

  const onResetPressed = () => {
    barricade?.resetRequestConfig();
    setRefreshList(Math.random());
  };

  const renderListItem = ({
    item,
    index,
  }: ListRenderItemInfo<RequestConfigForLib>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onListItemPressed(index)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        <Text style={[styles.value, themeColorStyle.textLight]}>
          {item.selectedResponseLabel}{' '}
          <Text style={[styles.icon, themeColorStyle.textLight]}>
            {Unicode.ChevronRight}
          </Text>
        </Text>
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{ title: Strings.Reset, onPress: onResetPressed }}
        title={Strings.Barricade}
        headerRight={{ title: Strings.Done, onPress: onDonePressed }}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={barricade?.requestConfig}
        renderItem={renderListItem}
        extraData={refreshList}
      />
      <Footer barricade={barricade} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: hScale(20),
    paddingVertical: vScale(12),
    borderBottomWidth: 1,
  },
  label: {
    fontSize: hScale(18),
    fontWeight: '400',
  },
  value: {
    fontSize: hScale(16),
  },
  icon: {
    fontSize: hScale(18),
  },
});

export { RequestList };
