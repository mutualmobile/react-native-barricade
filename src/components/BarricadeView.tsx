import React, { useState } from 'react';
import {
  Modal,
  NativeSyntheticEvent,
  SafeAreaView,
  StyleSheet,
} from 'react-native';

import { Barricade } from '../network';
import { ThemeProvider, ThemeType, useThemedColor } from '../theme';
import { RequestDetail } from './RequestDetail';
import { RequestList } from './RequestList';

enum ViewType {
  List,
  Detail,
}

export const BarricadeView = ({
  barricade,
  onRequestClose,
  theme = 'light',
  visible,
}: {
  barricade?: Barricade;
  onRequestClose?: (event: NativeSyntheticEvent<any>) => void;
  theme?: ThemeType;
  visible: boolean;
}) => {
  const { themeColorStyle } = useThemedColor();
  const [viewType, setViewType] = useState<ViewType>(ViewType.List);
  const [selectedListItemIndex, setSelectedListItemIndex] = useState<number>();

  const _onBackPressed = () => {
    setSelectedListItemIndex(undefined);
    setViewType(ViewType.List);
  };

  const _onListItemPressed = (index: number) => {
    setSelectedListItemIndex(index);
    setViewType(ViewType.Detail);
  };

  const renderContent = () => {
    if (viewType === ViewType.List) {
      return (
        <RequestList
          barricade={barricade}
          onListItemPressed={_onListItemPressed}
          onDonePressed={onRequestClose}
        />
      );
    } else {
      return (
        <RequestDetail
          barricade={barricade}
          selectedListItemIndex={selectedListItemIndex}
          onBackPressed={_onBackPressed}
        />
      );
    }
  };

  return (
    <ThemeProvider value={theme}>
      <Modal onRequestClose={onRequestClose} visible={visible}>
        <SafeAreaView
          style={[
            styles.container,
            themeColorStyle.background,
            themeColorStyle.border,
          ]}>
          {renderContent()}
        </SafeAreaView>
      </Modal>
    </ThemeProvider>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
