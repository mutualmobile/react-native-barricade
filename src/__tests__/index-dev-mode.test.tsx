/**
 * @jest-environment jsdom
 */
jest.useFakeTimers();
import React from 'react';
import { DeviceEventEmitter, DevSettings } from 'react-native';
import renderer, { act } from 'react-test-renderer';
import { EventType } from '../constants';

import {
  BarricadeView,
  createBarricade,
  getBarricadeInstance,
  showBarricadeView,
} from '../index';

describe('given that its __DEV__ mode', () => {
  describe('and barricade is undefined,', () => {
    test('should match snapshot', () => {
      const tree = renderer.create(<BarricadeView />);

      expect(tree).toMatchSnapshot();
      expect.assertions(1);
    });

    test('getBarricadeInstance should return undefined', () => {
      expect(getBarricadeInstance()).toBe(undefined);
      expect.assertions(1);
    });
  });

  test('and createBarricade is called, should create barricade instance and add Barricade to debug menu', () => {
    const spy = jest.spyOn(DevSettings, 'addMenuItem');

    const barricade = createBarricade();

    expect(barricade).not.toBe(undefined);
    expect(spy).toHaveBeenCalled();
    expect.assertions(2);
  });

  test('and showBarricadeView is called, should emit ShowBarricadeView event', () => {
    const spy = jest.spyOn(DeviceEventEmitter, 'emit');

    showBarricadeView();

    expect(spy).toHaveBeenCalledWith(EventType.ShowBarricadeView);
    expect.assertions(1);
  });

  describe('and barricade instance is created,', () => {
    let BarricadeInstance;
    beforeAll(() => {
      BarricadeInstance = createBarricade();
    });

    test('getBarricadeInstance should return Barricade instance', () => {
      expect(getBarricadeInstance()).toBe(BarricadeInstance);
      expect.assertions(1);
    });

    test('should add listener to ShowBarricadeView event when BarricadeView is mounted', () => {
      const addListenerSpy = jest.spyOn(DeviceEventEmitter, 'addListener');

      renderer.create(<BarricadeView />);

      expect(addListenerSpy).toHaveBeenCalledWith(
        EventType.ShowBarricadeView,
        expect.any(Function),
      );
      expect.assertions(1);
    });

    test('should set visibility of BarricadeView to true when showBarricadeView is called', async () => {
      const MMBarricadeView =
        require('../components/BarricadeView').BarricadeView;

      await act(async () => {
        const tree = renderer.create(<BarricadeView />);

        await jest.advanceTimersByTime(1000);
        showBarricadeView();
        await jest.advanceTimersByTime(1000);

        expect(tree.root.findByType(MMBarricadeView).props.visible).toBe(true);
      });

      expect.assertions(1);
    });

    test('should remove listener to ShowBarricadeView event when BarricadeView is unmounted', async () => {
      const removeAllListenersSpy = jest.spyOn(
        DeviceEventEmitter,
        'removeAllListeners',
      );

      const tree = renderer.create(<BarricadeView />);
      await act(async () => {
        tree.unmount();
      });

      expect(removeAllListenersSpy).toHaveBeenCalledWith(
        EventType.ShowBarricadeView,
      );
      expect.assertions(1);
    });

    test('should close the barricade view when BarricadeView calls onRequestClose', async () => {
      const MMBarricadeView =
        require('../components/BarricadeView').BarricadeView;

      await act(async () => {
        const tree = renderer.create(<BarricadeView />);

        await jest.advanceTimersByTime(1000);
        showBarricadeView();
        await jest.advanceTimersByTime(1000);

        // BarricadeView is visible
        expect(tree.root.findByType(MMBarricadeView).props.visible).toBe(true);

        tree.root.findByType(MMBarricadeView).props.onRequestClose();
        await jest.advanceTimersByTime(1000);

        // BarricadeView is not visible
        expect(tree.root.findByType(MMBarricadeView).props.visible).toBe(false);
      });

      expect.assertions(2);
    });
  });
});
