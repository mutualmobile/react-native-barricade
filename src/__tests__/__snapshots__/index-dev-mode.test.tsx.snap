// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given that its __DEV__ mode and barricade is undefined, should match snapshot 1`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  visible={false}
/>
`;
